src/main/java/com/google/devtools/build/lib/syntax/BinaryOperatorExpression.java

```diff
@@ -190,16 +190,16 @@ private static Object evaluate(
           return pipe(x, y, env, location);
 
         case AMPERSAND:
-          return (Integer) x & (Integer) y;
+          return and(x, y, location);
```

Comment:
I don't think function abstractions like 'and' help here. It makes the code longer unnecessarily, and more importantly it is the wrong control structure for allowing types (e.g. depset) to define their own operators. The code should first switch on the operator, then test supported pairs of operand types by cases (e.g. int+int, string+string), and then if none of those match, it should fall through to common code that tests whether the object defines its own `+` operator, and failing that, it should throw an exception ('unsupported operator: foo + bar').

Reply:
But again this would require refactoring this whole class as the same structure is followed for all other operators (a function that handles each operator). Or do you suggest following the structure you mentioned for the operators I added only?

Reply:
I confess I failed to notice that the existing cases already use functions, and only the previous draft of your `&` case did not. In that case feel free to follow the existing style and we can change all the cases in a later PR.  (The GitHub code review tool is far inferior to the internal tool we enjoy at Google, which makes it much easier to see the context.)